File: LeetCode/78.Subsets/Mario/78.Subsets.java
class Solution {

    private List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> subsets(int[] nums) {
        backtrack(nums, 0, new ArrayList<Integer>());
        return res;
    }

    private void backtrack(int[] nums, int index, List<Integer> array) {
        if (nums.length == index) {
            res.add(new ArrayList<>(array));
        } else {
            // not put num into array
            backtrack(nums, index + 1, array);

            // put num into array
            array.add(nums[index]);
            backtrack(nums, index + 1, array);
            array.remove(array.size() - 1);
        }
    }
}

File: LeetCode/78.Subsets/78.Subsets.swift
class Solution {
    func subsets(_ nums: [Int]) -> [[Int]] {
        let n = nums.count
        var res: [[Int]] = []

        func dfs(_ i: Int, _ path: [Int]) {
            guard i <= n else { return }
            res.append(path)

            for j in i ..< n {
                dfs(j + 1, path + [nums[j]])
            }
        }

        dfs(0, [])

        return res
    }
}
